### DISCUSSION TITLE
ALSA config on Ubuntu

### DISCUSSION THREAD

**Discussion by jbigdata-git** [mido/mido#312] [Q&A]
Hello,
I'm trying to use _mido_ on ubuntu with ALSA but don't pass a basic play.
I read the [doc](https://mido.readthedocs.io/en/latest/backends/rtmidi.html), here are the attempts I did but no one plays a sound.

> import mido
> import time
> import platform
> import os
> 
> print('platform:', platform.platform(), os.name)
> print('mido infos:', mido.backend.module.get_api_names(), mido.get_output_names())
> #mido.set_backend('mido.backends.pygame')
> #mido.set_backend('mido.backends.rtmidi')
> #output = mido.open_output('New Port', virtual=True)
> #output = mido.open_output()
> #output = mido.open_output('Midi Through Port-0')
> mido.set_backend('mido.backends.rtmidi/LINUX_ALSA')
> output = mido.open_output()
> print('output port:', output)
> msg = mido.Message('note_on', note=64, velocity=60)
> output.send(msg)
> time.sleep(2)
> output.close()

I also try a midi file

> ...
> mido.MidiFile('../_midi/tuto_drumsnotes.mid', clip=True)
> for msg in mid.play():
>   port.send(msg)
> ...

How use LINUX_ALSA ?

JB.

**Comment by olemb** [comment]
Hi,

I'm also running on Linux/ALSA. I'd like it to work with ALSA out of the box
but for now you need to install the python-rtmidi package from PyPI.

Let me know if you have any questions.

- Ole

**Reply by jbigdata-git** [reply]
Thanks for your answer.
I have already installed _python-rtmidi_.
_Requirement already satisfied: python-rtmidi in /usr/local/lib/python3.8/dist-packages (1.4.9)_

To help I can add that "my" ALSA (with _pulseaudio_) sounds fine with _fluidsynth_, _pygame_, _timidity_...

**Comment by olemb** [comment]
Can you try this and see if you get the same result?

```
$ python3
Python 3.8.6 (default, Sep 25 2020, 09:36:53)
[GCC 10.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import mido
>>> mido.get_input_names()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/olemb/src/mido/mido/backends/backend.py", line 169, in
get_input_names
    devices = self._get_devices(**self._add_api(kwargs))
  File "/home/olemb/src/mido/mido/backends/backend.py", line 162, in
_get_devices
    if hasattr(self.module, 'get_devices'):
  File "/home/olemb/src/mido/mido/backends/backend.py", line 42, in module
    self.load()
  File "/home/olemb/src/mido/mido/backends/backend.py", line 58, in load
    self._module = importlib.import_module(self.name)
  File "/usr/lib/python3.8/importlib/__init__.py", line 127, in
import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 783, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in
_call_with_frames_removed
  File "/home/olemb/src/mido/mido/backends/rtmidi.py", line 8, in <module>
    import rtmidi
ModuleNotFoundError: No module named 'rtmidi'
>>> import rtmidi
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'rtmidi'
```

**Comment by olemb** [comment]
Aha, I think I've figured it out. There's an extra ")" on this line:

    mido.get_output_names())

What happens if you remove it?

**Reply by jbigdata-git** [reply]
I think this problem is more tricky than a compile typo.
extra ")" comes with _print_ "(". The test in the first post executes but there's no sound.

Here's the output of _mido_ internals:

_$ uname -a_
Linux vm0 5.4.0-73-generic #82-Ubuntu SMP Wed Apr 14 17:39:42 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
_$ python3_
Python 3.8.5 (default, Jan 27 2021, 15:41:15)
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
`>>>` _import mido_
`>>>` _import rtmidi_
`>>>` _mido.get_output_names()_
**['Midi Through:Midi Through Port-0 14:0', 'Midi Through:Midi Through Port-0 14:0']**
`>>>` _mido.get_input_names()_
**['Midi Through:Midi Through Port-0 14:0', 'Midi Through:Midi Through Port-0 14:0']**
`>>>` _mido.backend.module.get_api_names()_
**['LINUX_ALSA', 'UNIX_JACK']**

To investigate, can U share the same proper output done on your platform ?

**Comment by olemb** [comment]
I think we introdced a bug in 1.2.10 when we removed the sorting of ports.
(Note that both inputs and both outputs have the same name. This was not
the case in 1.2.9.)

Definitely something that need to be fixed. Could you open an issue on this
so we can discuss it in public?

**Comment by olemb** [comment]
On the other hand, there is no need for an issue since we are already
discussing it here in this email thread.

I don't have access to GitHub from this computer at the moment but I'll
start debugging today and add an issue later if needed.

**Reply by jbigdata-git** [reply]
For my information, can U even share your output.
So I wait for a patch or workaround to keep on testing _mido_.  

Thanks for your support.

**Comment by olemb** [comment]
I'll share my output. Just don't have things set up yet on my new computer.
Need to install a few things.

**Comment by olemb** [comment]
Thank you for helping improve Mido!
